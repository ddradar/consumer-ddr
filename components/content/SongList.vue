<template>
  <OTable :data="songs!" striped narrowed :loading="isLoading">
    <OTableColumn v-slot="props" field="name" label="Name">
      <NuxtLink :to="`/songs/${props.row.slug}/`">
        {{ props.row.name }}
      </NuxtLink>
    </OTableColumn>
    <OTableColumn v-slot="props" field="artist" label="Artist">
      <span>{{ props.row.artist }}</span>
    </OTableColumn>
    <OTableColumn v-slot="props" field="bpm" label="BPM">
      {{ props.row.bpm }}
    </OTableColumn>
    <OTableColumn
      v-for="playStyle in playStyles"
      v-slot="props"
      :key="playStyle"
      :label="playStyle"
    >
      <div class="tags has-addons">
        <span
          v-for="chart in props.row.charts[playStyle]"
          :key="chart.difficulty"
          class="tag"
          :class="chart.color"
          :title="chart.name"
        >
          {{ chart.level }}
        </span>
      </div>
    </OTableColumn>

    <template #empty>
      <section class="section">
        <div class="content has-text-grey has-text-centered">
          <p>Nothing here.</p>
        </div>
      </section>
    </template>
  </OTable>
</template>

<script lang="ts" setup>
import type {
  Chart,
  PlayStyle,
  SoftwareParsedContent,
  SongParsedContent
} from '~~/src/content'
import { normalizeDifficulty } from '~~/src/song'

type ChartInfo = Pick<Chart, 'level' | 'difficulty'> & {
  color: string
  name: string
}
interface Props {
  series: string
}

const prop = defineProps<Props>()

const { data: software } = await useAsyncData(`/software/${prop.series}`, () =>
  queryContent<SoftwareParsedContent>(prop.series)
    .where({ _type: 'markdown' })
    .findOne()
)

const { data: songs, pending: isLoading } = await useAsyncData(
  `/software/${prop.series}/songs`,
  () =>
    queryContent<SongParsedContent>(prop.series)
      .where({ _type: 'json' })
      .without('series')
      .find(),
  {
    transform: (songs: SongParsedContent[]) =>
      songs.map((s) => ({
        slug: s.slug,
        name: s.name,
        artist: s.artist,
        bpm: s.bpm,
        charts: s.charts.reduce((p, c) => {
          if (!p[c.playStyle]) {
            p[c.playStyle] = []
          }
          const name =
            software.value?.difficultyNames[c.difficulty] ?? 'Unknown'
          p[c.playStyle]!.push({
            level: c.level,
            difficulty: c.difficulty,
            name,
            color: `is-${normalizeDifficulty(name).toLowerCase()}`
          })
          return p
        }, {} as Partial<Record<PlayStyle, ChartInfo[]>>)
      }))
  }
)

const playStyles = computed(() => [
  ...new Set(songs.value?.flatMap((s) => Object.keys(s.charts)))
])
</script>
