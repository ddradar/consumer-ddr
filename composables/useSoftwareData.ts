import type { ParsedContent } from '@nuxt/content/dist/runtime/types'
import type { Ref } from 'vue'

import type seriesList from '~~/content/series/.seriesList.json'

export const regionList = ['JP', 'US', 'EU', 'None'] as const

export const platforms = [
  'Play Station',
  'Play Station 2',
  'Dreamcast',
  'GAME BOY COLOR',
  'GAME CUBE',
  'Wii',
  'Xbox 360',
  'Windows',
] as const

export interface Software {
  slug: keyof typeof seriesList
  title: string
  platform: (typeof platforms)[number]
  region: (typeof regionList)[number]
  launched: string
  difficulties: { [key in number]: { name: string; class: string } }
}

export default async function (slug: string) {
  const { data } = await useAsyncData(`/software/${slug}`, () =>
    queryContent<Software & ParsedContent>()
      .where({ _type: 'markdown', slug })
      .findOne()
  )

  return { software: data as Ref<Software & ParsedContent> }
}
