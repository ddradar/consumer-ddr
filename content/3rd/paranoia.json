{
  "series": "3rd",
  "name": "PARANOiA",
  "artist": "180",
  "bpm": 180,
  "charts": [
    {
      "playStyle": "SINGLE",
      "difficulty": 1,
      "level": 6
    },
    {
      "playStyle": "SINGLE",
      "difficulty": 2,
      "level": 7
    },
    {
      "playStyle": "SINGLE",
      "difficulty": 3,
      "level": 8
    },
    {
      "playStyle": "DOUBLE",
      "difficulty": 1,
      "level": 7
    },
    {
      "playStyle": "DOUBLE",
      "difficulty": 2,
      "level": 8
    },
    {
      "playStyle": "UNISON",
      "difficulty": 1,
      "level": 6
    },
    {
      "playStyle": "UNISON",
      "difficulty": 2,
      "level": 7
    },
    {
      "playStyle": "UNISON",
      "difficulty": 3,
      "level": 8
    }
  ]
}
