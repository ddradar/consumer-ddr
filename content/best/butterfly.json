{
  "slug": "butterfly",
  "series": "best",
  "name": "butterfly",
  "artist": "smile. dk",
  "bpm": 135,
  "charts": [
    {
      "playStyle": "SINGLE",
      "difficulty": 0,
      "level": 1
    },
    {
      "playStyle": "SINGLE",
      "difficulty": 1,
      "level": 3
    },
    {
      "playStyle": "SINGLE",
      "difficulty": 2,
      "level": 4
    },
    {
      "playStyle": "SINGLE",
      "difficulty": 3,
      "level": 5
    },
    {
      "playStyle": "DOUBLE",
      "difficulty": 1,
      "level": 4
    },
    {
      "playStyle": "DOUBLE",
      "difficulty": 2,
      "level": 5
    },
    {
      "playStyle": "DOUBLE",
      "difficulty": 3,
      "level": 6
    },
    {
      "playStyle": "UNISON",
      "difficulty": 1,
      "level": 3
    },
    {
      "playStyle": "UNISON",
      "difficulty": 2,
      "level": 4
    },
    {
      "playStyle": "UNISON",
      "difficulty": 3,
      "level": 5
    }
  ]
}
