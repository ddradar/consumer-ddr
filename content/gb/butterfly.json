{
  "slug": "butterfly",
  "series": "gb",
  "name": "butterfly",
  "artist": "smile. dk",
  "bpm": 135,
  "charts": [
    {
      "playStyle": "SINGLE",
      "difficulty": 5,
      "level": 3
    },
    {
      "playStyle": "SINGLE",
      "difficulty": 1,
      "level": 3
    },
    {
      "playStyle": "SINGLE",
      "difficulty": 2,
      "level": 4
    },
    {
      "playStyle": "SINGLE",
      "difficulty": 3,
      "level": 6
    },
    {
      "playStyle": "COUPLE",
      "difficulty": 5,
      "level": 3
    },
    {
      "playStyle": "COUPLE",
      "difficulty": 1,
      "level": 3
    },
    {
      "playStyle": "COUPLE",
      "difficulty": 2,
      "level": 4
    },
    {
      "playStyle": "COUPLE",
      "difficulty": 3,
      "level": 6
    }
  ]
}
