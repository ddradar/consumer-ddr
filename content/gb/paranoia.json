{
  "slug": "paranoia",
  "series": "gb",
  "name": "PARANOiA",
  "artist": "180",
  "bpm": 180,
  "charts": [
    {
      "playStyle": "SINGLE",
      "difficulty": 1,
      "level": 6
    },
    {
      "playStyle": "SINGLE",
      "difficulty": 2,
      "level": 7
    },
    {
      "playStyle": "SINGLE",
      "difficulty": 3,
      "level": 8
    },
    {
      "playStyle": "COUPLE",
      "difficulty": 1,
      "level": 6
    },
    {
      "playStyle": "COUPLE",
      "difficulty": 2,
      "level": 7
    },
    {
      "playStyle": "COUPLE",
      "difficulty": 3,
      "level": 8
    }
  ]
}
