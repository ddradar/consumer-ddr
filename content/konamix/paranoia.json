{
  "slug": "paranoia",
  "series": "konamix",
  "name": "PARANOiA",
  "artist": "180",
  "bpm": 180,
  "charts": [
    { "playStyle": "SINGLE", "difficulty": 1, "level": 6 },
    { "playStyle": "SINGLE", "difficulty": 2, "level": 7 },
    { "playStyle": "SINGLE", "difficulty": 3, "level": 8 },
    { "playStyle": "DOUBLE", "difficulty": 1, "level": 7 },
    { "playStyle": "DOUBLE", "difficulty": 2, "level": 8 },
    { "playStyle": "DOUBLE", "difficulty": 3, "level": 9 },
    { "playStyle": "6-PANELS", "difficulty": 1, "level": 6 },
    { "playStyle": "6-PANELS", "difficulty": 2, "level": 7 },
    { "playStyle": "6-PANELS", "difficulty": 3, "level": 8 },
    { "playStyle": "BATTLE", "difficulty": 1, "level": 7 }
  ]
}
