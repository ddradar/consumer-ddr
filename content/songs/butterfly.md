---
slug: "butterfly"
name: "butterfly"
artist: "smile. dk"
bpm: 135
series:
  - "1st-jp"
  - "2nd-dc"
  - "best"
  - "gb"
  - "party"
---

# {{ $doc.name }} / {{ $doc.artist }}

|Artist|BPM|
|------|---|
|{{ $doc.artist }}|{{ $doc.bpm }}|

## 収録タイトル

- [Dance Dance Revolution (JP)](/series/1st-jp/)
- [Dance Dance Revolution 2nd MIX Dreamcast Edition](/series/2nd-dc/)
- [Dance Dance Revolution BEST HITS](/series/best/)
- [Dance Dance Revolution GB](/series/gb/)
- [Dance Dance Revolution Party Collection](/series/party/)

## 譜面一覧

|PlayStyle|Difficulty|Levels|Movie|
|---------|----------|------|-----|
|SINGLE|[BEGINNER]{.tag.is-beginner}|:level{slug="party" level=1}||
|SINGLE|[BEGINNER]{.tag.is-beginner}|:level{slug="best" level=1}||
|SINGLE|[NORMAL]{.tag.is-basic}|:level{slug="1st-jp" level=3} :level{slug="gb" level=2}||
|SINGLE|[BASIC]{.tag.is-basic}|:level{slug="1st-jp" level=3} :level{slug="2nd-dc" level=3} :level{slug="best" level=3} :level{slug="gb" level=3} :level{slug="party" level=3}||
|SINGLE|[DIFFICULT]{.tag.is-difficult}|:level{slug="1st-jp" level=4} :level{slug="2nd-dc" level=4} :level{slug="best" level=4} :level{slug="gb" level=4} :level{slug="party" level=4}||
|SINGLE|[EXPERT]{.tag.is-expert}|:level{slug="1st-jp" level=6} :level{slug="2nd-dc" level=5} :level{slug="best" level=5} :level{slug="gb" level=6} :level{slug="party" level=5}||
|SINGLE|[CHALLENGE]{.tag.is-challenge}|:level{slug="party" level=8}||
|DOUBLE|[BASIC]{.tag.is-basic}|:level{slug="1st-jp" level=4} :level{slug="2nd-dc" level=4} :level{slug="best" level=4} :level{slug="party" level=4}||
|DOUBLE|[DIFFICULT]{.tag.is-difficult}|:level{slug="1st-jp" level=6} :level{slug="2nd-dc" level=5} :level{slug="best" level=5} :level{slug="party" level=5}||
|DOUBLE|[EXPERT]{.tag.is-expert}|:level{slug="best" level=6} :level{slug="party" level=6}||
|DOUBLE|[CHALLENGE]{.tag.is-challenge}|:level{slug="party" level=8}||
|COUPLE|[NORMAL]{.tag.is-basic}|:level{slug="1st-jp" level=3} :level{slug="gb" level=2}||
|COUPLE|[BASIC]{.tag.is-basic}|:level{slug="1st-jp" level=3} :level{slug="2nd-dc" level=3} :level{slug="gb" level=3}||
|COUPLE|[DIFFICULT]{.tag.is-difficult}|:level{slug="1st-jp" level=4} :level{slug="2nd-dc" level=4} :level{slug="gb" level=4}||
|COUPLE|[EXPERT]{.tag.is-expert}|:level{slug="gb" level=3} :level{slug="1st-jp" level=6} :level{slug="2nd-dc" level=5} :level{slug="gb" level=5}|
|UNISON|[BASIC]{.tag.is-basic}|:level{slug="best" level=3}||
|UNISON|[DIFFICULT]{.tag.is-difficult}|:level{slug="best" level=4}||
|UNISON|[EXPERT]{.tag.is-expert}|:level{slug="best" level=5}||
