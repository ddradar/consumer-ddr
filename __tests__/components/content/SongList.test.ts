import { RouterLinkStub } from '@vue/test-utils'
import { describe, expect, test, vi } from 'vitest'
import { ref } from 'vue'

import SongList from '~~/components/content/SongList.vue'
import useSoftwareData from '~~/composables/useSoftwareData'
import butterfly from '~~/content/1st-jp/butterfly.json'
import paranoia from '~~/content/1st-jp/paranoia.json'

import { mockSoftware, mountAsync, plugins } from '../../test-utils'

vi.mock('~~/composables/useSoftwareData')

const rawSongData = [butterfly, paranoia].map((d) => ({
  slug: d.slug,
  name: d.name,
  artist: d.artist,
  bpm: d.bpm,
  charts: d.charts
}))

describe('components/content/SongList.vue', () => {
  const songs = [
    {
      slug: butterfly.slug,
      name: butterfly.name,
      artist: butterfly.artist,
      bpm: butterfly.bpm,
      charts: {
        SINGLE: [
          { difficulty: 6, level: 3, name: 'NORMAL', color: 'is-basic' },
          { difficulty: 1, level: 3, name: 'BASIC', color: 'is-basic' },
          { difficulty: 2, level: 4, name: 'ANOTHER', color: 'is-difficult' },
          { difficulty: 3, level: 6, name: 'MANIAC', color: 'is-expert' }
        ],
        DOUBLE: [
          { difficulty: 1, level: 4, name: 'BASIC', color: 'is-basic' },
          { difficulty: 2, level: 6, name: 'ANOTHER', color: 'is-difficult' }
        ],
        COUPLE: [
          { difficulty: 6, level: 3, name: 'NORMAL', color: 'is-basic' },
          { difficulty: 1, level: 3, name: 'BASIC', color: 'is-basic' },
          { difficulty: 2, level: 4, name: 'ANOTHER', color: 'is-difficult' },
          { difficulty: 3, level: 6, name: 'MANIAC', color: 'is-expert' }
        ]
      }
    },
    {
      slug: paranoia.slug,
      name: paranoia.name,
      artist: paranoia.artist,
      bpm: paranoia.bpm,
      charts: {
        SINGLE: [
          { difficulty: 1, level: 6, name: 'BASIC', color: 'is-basic' },
          { difficulty: 2, level: 7, name: 'ANOTHER', color: 'is-difficult' },
          { difficulty: 3, level: 8, name: 'MANIAC', color: 'is-expert' }
        ],
        DOUBLE: [
          { difficulty: 1, level: 7, name: 'BASIC', color: 'is-basic' },
          { difficulty: 2, level: 8, name: 'ANOTHER', color: 'is-difficult' }
        ],
        COUPLE: [
          { difficulty: 1, level: 6, name: 'BASIC', color: 'is-basic' },
          { difficulty: 2, level: 7, name: 'ANOTHER', color: 'is-difficult' },
          { difficulty: 3, level: 8, name: 'MANIAC', color: 'is-expert' }
        ]
      }
    }
  ]
  const global = { plugins, stubs: { NuxtLink: RouterLinkStub } }
  const options = { global, props: { series: mockSoftware.slug } }

  describe('snapshot test', () => {
    test('renders loading state', async () => {
      // Arrange
      vi.mocked(useSoftwareData).mockResolvedValue({
        software: ref(mockSoftware)
      })
      vi.mocked(useAsyncData).mockResolvedValue({
        data: ref([]),
        pending: ref(true)
      } as any)

      // Act
      const wrapper = await mountAsync(SongList, options)

      // Assert
      expect(wrapper.element).toMatchSnapshot()
    })

    test('renders song info', async () => {
      // Arrange
      vi.mocked(useSoftwareData).mockResolvedValue({
        software: ref(mockSoftware)
      })
      vi.mocked(useAsyncData).mockResolvedValue({
        data: ref(songs),
        pending: ref(false)
      } as any)

      // Act
      const wrapper = await mountAsync(SongList, options)

      // Assert
      expect(wrapper.element).toMatchSnapshot()
    })
  })

  describe('useAsyncData()', () => {
    test('calls with transform option', async () => {
      // Arrange
      vi.mocked(useSoftwareData).mockResolvedValue({
        software: ref(mockSoftware)
      })
      vi.mocked(useAsyncData).mockReset()
      vi.mocked(useAsyncData).mockResolvedValue({
        data: ref([]),
        pending: ref(true)
      } as any)

      // Act
      await mountAsync(SongList, options)
      const transform = vi.mocked(useAsyncData).mock.calls[0][2]!.transform!

      // Assert
      expect(transform(rawSongData)).toStrictEqual(songs)
    })
  })
})
