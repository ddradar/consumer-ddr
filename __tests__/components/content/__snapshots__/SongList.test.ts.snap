// Vitest Snapshot v1

exports[`components/content/SongList.vue > snapshot test > renders loading state 1`] = `
<div
  class="b-table"
>
  <div
    style="display: none;"
  >
    
    <span
      data-id="1"
    >
      Name
    </span>
    <span
      data-id="2"
    >
      Artist
    </span>
    <span
      data-id="3"
    >
      BPM
    </span>
    
    
    
  </div>
  <!--v-if-->
  <!--v-if-->
  <div
    class="table-wrapper"
  >
    <table
      class="table is-striped is-narrow is-empty"
      tabindex="false"
    >
      <!--v-if-->
      <thead>
        <tr>
          <!--v-if-->
          <!--v-if-->
          
          <th
            class=""
            draggable="false"
          >
            <span>
              Name 
              <span
                class="th-wrap sort-icon"
                style="display: none;"
              >
                <span
                  class="icon is-small"
                >
                  <i
                    class="mdi mdi-arrow-up"
                  />
                </span>
              </span>
            </span>
          </th>
          <th
            class=""
            draggable="false"
          >
            <span>
              Artist 
              <span
                class="th-wrap sort-icon"
                style="display: none;"
              >
                <span
                  class="icon is-small"
                >
                  <i
                    class="mdi mdi-arrow-up"
                  />
                </span>
              </span>
            </span>
          </th>
          <th
            class=""
            draggable="false"
          >
            <span>
              BPM 
              <span
                class="th-wrap sort-icon"
                style="display: none;"
              >
                <span
                  class="icon is-small"
                >
                  <i
                    class="mdi mdi-arrow-up"
                  />
                </span>
              </span>
            </span>
          </th>
          
          <!--v-if-->
        </tr>
        <!--v-if-->
        <!--v-if-->
      </thead>
      <tbody>
        
        
        <tr>
          <td
            colspan="3"
          >
            
            <section
              class="section"
            >
              <div
                class="content has-text-grey has-text-centered"
              >
                <p>
                  Nothing here.
                </p>
              </div>
            </section>
            
          </td>
        </tr>
      </tbody>
      <!--v-if-->
    </table>
    
    <transition-stub>
      <div
        class="o-load loading"
      >
        <div
          class="o-load__overlay loading-overlay"
        />
        
        <span
          class="icon is-spin is-medium o-load__icon icon"
        >
          <i
            class="mdi mdi-loading mdi-36px"
          />
        </span>
        
      </div>
    </transition-stub>
    
  </div>
  <!--v-if-->
</div>
`;

exports[`components/content/SongList.vue > snapshot test > renders song info 1`] = `
<div
  class="b-table"
>
  <div
    style="display: none;"
  >
    
    <span
      data-id="1"
    >
      Name
    </span>
    <span
      data-id="2"
    >
      Artist
    </span>
    <span
      data-id="3"
    >
      BPM
    </span>
    
    <span
      data-id="4"
    >
      SINGLE
    </span>
    <span
      data-id="5"
    >
      DOUBLE
    </span>
    <span
      data-id="6"
    >
      COUPLE
    </span>
    
    
  </div>
  <!--v-if-->
  <!--v-if-->
  <div
    class="table-wrapper"
  >
    <table
      class="table is-striped is-narrow"
      tabindex="false"
    >
      <!--v-if-->
      <thead>
        <tr>
          <!--v-if-->
          <!--v-if-->
          
          <th
            class=""
            draggable="false"
          >
            <span>
              Name 
              <span
                class="th-wrap sort-icon"
                style="display: none;"
              >
                <span
                  class="icon is-small"
                >
                  <i
                    class="mdi mdi-arrow-up"
                  />
                </span>
              </span>
            </span>
          </th>
          <th
            class=""
            draggable="false"
          >
            <span>
              Artist 
              <span
                class="th-wrap sort-icon"
                style="display: none;"
              >
                <span
                  class="icon is-small"
                >
                  <i
                    class="mdi mdi-arrow-up"
                  />
                </span>
              </span>
            </span>
          </th>
          <th
            class=""
            draggable="false"
          >
            <span>
              BPM 
              <span
                class="th-wrap sort-icon"
                style="display: none;"
              >
                <span
                  class="icon is-small"
                >
                  <i
                    class="mdi mdi-arrow-up"
                  />
                </span>
              </span>
            </span>
          </th>
          <th
            class=""
            draggable="false"
          >
            <span>
              SINGLE 
              <span
                class="th-wrap sort-icon"
                style="display: none;"
              >
                <span
                  class="icon is-small"
                >
                  <i
                    class="mdi mdi-arrow-up"
                  />
                </span>
              </span>
            </span>
          </th>
          <th
            class=""
            draggable="false"
          >
            <span>
              DOUBLE 
              <span
                class="th-wrap sort-icon"
                style="display: none;"
              >
                <span
                  class="icon is-small"
                >
                  <i
                    class="mdi mdi-arrow-up"
                  />
                </span>
              </span>
            </span>
          </th>
          <th
            class=""
            draggable="false"
          >
            <span>
              COUPLE 
              <span
                class="th-wrap sort-icon"
                style="display: none;"
              >
                <span
                  class="icon is-small"
                >
                  <i
                    class="mdi mdi-arrow-up"
                  />
                </span>
              </span>
            </span>
          </th>
          
          <!--v-if-->
        </tr>
        <!--v-if-->
        <!--v-if-->
      </thead>
      <tbody>
        
        
        <tr
          class=""
          draggable="false"
        >
          <!--v-if-->
          <!--v-if-->
          
          <td
            class=""
            data-label="Name"
          >
            <a>
              butterfly
            </a>
          </td>
          <td
            class=""
            data-label="Artist"
          >
            <span>
              smile. dk
            </span>
          </td>
          <td
            class=""
            data-label="BPM"
          >
            135
          </td>
          <td
            class=""
            data-label="SINGLE"
          >
            <div
              class="tags has-addons"
            >
              
              <span
                class="tag is-basic"
                title="NORMAL"
              >
                3
              </span>
              <span
                class="tag is-basic"
                title="BASIC"
              >
                3
              </span>
              <span
                class="tag is-difficult"
                title="ANOTHER"
              >
                4
              </span>
              <span
                class="tag is-expert"
                title="MANIAC"
              >
                6
              </span>
              
            </div>
          </td>
          <td
            class=""
            data-label="DOUBLE"
          >
            <div
              class="tags has-addons"
            >
              
              <span
                class="tag is-basic"
                title="BASIC"
              >
                4
              </span>
              <span
                class="tag is-difficult"
                title="ANOTHER"
              >
                6
              </span>
              
            </div>
          </td>
          <td
            class=""
            data-label="COUPLE"
          >
            <div
              class="tags has-addons"
            >
              
              <span
                class="tag is-basic"
                title="NORMAL"
              >
                3
              </span>
              <span
                class="tag is-basic"
                title="BASIC"
              >
                3
              </span>
              <span
                class="tag is-difficult"
                title="ANOTHER"
              >
                4
              </span>
              <span
                class="tag is-expert"
                title="MANIAC"
              >
                6
              </span>
              
            </div>
          </td>
          
          <!--v-if-->
        </tr>
        <transition-stub>
          <!--v-if-->
        </transition-stub>
        <!--v-if-->
        
        
        <tr
          class=""
          draggable="false"
        >
          <!--v-if-->
          <!--v-if-->
          
          <td
            class=""
            data-label="Name"
          >
            <a>
              PARANOiA
            </a>
          </td>
          <td
            class=""
            data-label="Artist"
          >
            <span>
              180
            </span>
          </td>
          <td
            class=""
            data-label="BPM"
          >
            180
          </td>
          <td
            class=""
            data-label="SINGLE"
          >
            <div
              class="tags has-addons"
            >
              
              <span
                class="tag is-basic"
                title="BASIC"
              >
                6
              </span>
              <span
                class="tag is-difficult"
                title="ANOTHER"
              >
                7
              </span>
              <span
                class="tag is-expert"
                title="MANIAC"
              >
                8
              </span>
              
            </div>
          </td>
          <td
            class=""
            data-label="DOUBLE"
          >
            <div
              class="tags has-addons"
            >
              
              <span
                class="tag is-basic"
                title="BASIC"
              >
                7
              </span>
              <span
                class="tag is-difficult"
                title="ANOTHER"
              >
                8
              </span>
              
            </div>
          </td>
          <td
            class=""
            data-label="COUPLE"
          >
            <div
              class="tags has-addons"
            >
              
              <span
                class="tag is-basic"
                title="BASIC"
              >
                6
              </span>
              <span
                class="tag is-difficult"
                title="ANOTHER"
              >
                7
              </span>
              <span
                class="tag is-expert"
                title="MANIAC"
              >
                8
              </span>
              
            </div>
          </td>
          
          <!--v-if-->
        </tr>
        <transition-stub>
          <!--v-if-->
        </transition-stub>
        <!--v-if-->
        
        
        <!--v-if-->
      </tbody>
      <!--v-if-->
    </table>
    
    <transition-stub>
      <!--v-if-->
    </transition-stub>
    
  </div>
  <!--v-if-->
</div>
`;
